### PR TITLE
TuningStage refactoring

### PR DESCRIPTION
**Pull request by jana-d**
This is ongoing work refactoring all TuningStage classes. Currently, changes were made mainly in the TuningStage base class, moving the original methods as functions into base_tasks.py and take_data.py.

This draft pull request is intended to start a discussion on the implementation of these functions and how to make them flexible enough so that they can be used in a non-nanotune setting. 
While some function do depend on the nt.Gate abstraction, this will change once nanotune switches to a common device abstraction. 
@guenp any thoughts on functions in base_tasks.py and take_data.py.?

Documentation of the initial classes is yet to be finished. The current priority is to fix those portions of code which will be used first.

Tasks/functions for gate characterizations and charge diagrams will follow next, in the same style as the base tasks if it (the style) is deemed suitable.
(Typing will be improved by adding aliases.... )

### CODE REVIEW

File: nanotune/tuningstages/tuningstage.py [hunk 1]
```diff
@@ -108,41 +126,46 @@ def fit_class(self):
...
-        pass
+    def conclude_iteration(
+        self,
+        tuning_result: TuningResult,
+        voltage_ranges: List[Tuple[float, float]],
+        safety_voltage_ranges: List[Tuple[float, float]],
+        count: int,
+        n_iterations: int,
+    ) -> Tuple[bool, List[Tuple[float, float]], List[str]]:
+        """ """
```
Comment (guenp) [inline nanotune/tuningstages/tuningstage.py:137]: Is something missing here?

File: nanotune/tuningstages/tuningstage.py [hunk 2]
```diff
@@ -108,41 +126,46 @@ def fit_class(self):
...
     @abstractmethod
-    def update_current_ranges(
+    def verify_classification_result(
         self,
-        actions: List[str],
-    ) -> None:
-        """"""
+        ml_result: Dict[str, int],
+    ) -> bool:
+        """ """
```
Comment (guenp) [inline nanotune/tuningstages/tuningstage.py:144]: Same here, is this a placeholder?

**Comment by guenp** [comment]
I want to take a more in-depth look at this to give more useful feedback but so far I think you've done a great job splitting this class up into building blocks. I really like how you are passing a list of tasks to a common method that runs them in sequence.

Just a few questions:
- What are `range_update_directives`?
- Why is ramping deactivated before measurements?
- I'm seeing a lot of missing docstrings, would you mind just adding a sentence or so just to explain what these methods/functions do?

**Comment by jana-d** [comment]
@guenp Yes, the initial nanotune classes still need to be documented. I was focusing on the functions which would be used in a non-nanotune setting. Apologies for that, I will do that today.
My suggestion would be to look at base_tasks.py and take_data.py to see if the refactoring is what you had in mind. 

`range_update_directives` is a list of strings indicating if the current voltage ranges need to be extended or shifted. It is useful to detect completely pinched-off or open regimes. It gives an indication for either further tuning steps on how to proceed or can be used to update the current ranges and re-take the same measurement. It is an option not used in the `run_stage()` function, which executed typical tuning stage tasks. In case one would like to use `range_update_directives`, then 'iterate_stage()' is the function to use. It wraps `run_stage()` and updated voltage ranges if needed.

I was setting off ramping during measurements to save time. We can have also a basic `take_data` without this functionality if we prefer - or simple not use the generator taking care of switching off the ramping.

**Comment by guenp** [comment]
> My suggestion would be to look at base_tasks.py and take_data.py to see if the refactoring is what you had in mind.

That sounds good! It is indeed what I had in mind. 👍 I think we can think about the exact arguments being passed around and making sure they're serializable but that is a push we can do later.

> range_update_directives is a list of strings indicating if the current voltage ranges need to be extended or shifted. 

Currently they are freeform strings which is a great start but if we want it to be a little more robust it might be good thing to make this a little more structured, i.e. create a dedicated class or `Enum` to describe this feature.

> I was setting off ramping during measurements to save time.

Was this time saving meant for the actual measurement or for just the unit tests/simulation?

**Comment by jana-d** [comment]
Perfect! Thank you for having a look! :)

> I think we can think about the exact arguments being passed around and making sure they're serializable but that is a push we can do later.

Agreed!

> if we want it to be a little more robust it might be good thing to make this a little more structured, i.e. create a dedicated class or `Enum` to describe this feature.

Agreed!

> Was this time saving meant for the actual measurement or for just the unit tests/simulation?

It was to save time during measurements.

**Pull request merged**